File: employee-management-api/src/main/java/com/ems/service/FormFieldConfigService.java
package com.ems.service;

import com.ems.dto.FormFieldConfigDTO;
import com.ems.exception.ResourceNotFoundException;
import com.ems.model.FormFieldConfig;
import com.ems.repository.FormFieldConfigRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
public class FormFieldConfigService {

    private final FormFieldConfigRepository repository;

    @Transactional(readOnly = true)
    public List<FormFieldConfigDTO> getAllConfigs() {
        return repository.findAllByOrderByTabNameAscSortOrderAscIdAsc()
                .stream()
                .map(FormFieldConfigDTO::fromEntity)
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public List<FormFieldConfigDTO> getVisibleConfigs() {
        return repository.findByIsVisibleTrueOrderByTabNameAscSortOrderAscIdAsc()
                .stream()
                .map(FormFieldConfigDTO::fromEntity)
                .collect(Collectors.toList());
    }

    @Transactional
    public FormFieldConfigDTO toggleMandatory(Long id) {
        FormFieldConfig config = repository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("FormFieldConfig not found with id: " + id));
        config.setIsMandatory(!Boolean.TRUE.equals(config.getIsMandatory()));
        return FormFieldConfigDTO.fromEntity(repository.save(config));
    }

    @Transactional
    public FormFieldConfigDTO toggleVisibility(Long id) {
        FormFieldConfig config = repository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("FormFieldConfig not found with id: " + id));
        config.setIsVisible(!Boolean.TRUE.equals(config.getIsVisible()));
        return FormFieldConfigDTO.fromEntity(repository.save(config));
    }

    @Transactional
    public List<FormFieldConfigDTO> updateBulk(List<FormFieldConfigDTO> dtos) {
        for (FormFieldConfigDTO dto : dtos) {
            if (dto.getId() != null) {
                repository.findById(dto.getId()).ifPresent(entity -> {
                    if (dto.getIsMandatory() != null) entity.setIsMandatory(dto.getIsMandatory());
                    if (dto.getIsVisible() != null) entity.setIsVisible(dto.getIsVisible());
                    if (dto.getFieldLabel() != null) entity.setFieldLabel(dto.getFieldLabel());
                    if (dto.getTabName() != null) entity.setTabName(dto.getTabName());
                    if (dto.getSortOrder() != null) entity.setSortOrder(dto.getSortOrder());
                    if (dto.getPlaceholder() != null) entity.setPlaceholder(dto.getPlaceholder());
                    repository.save(entity);
                });
            }
        }
        return getAllConfigs();
    }

    @Transactional
    public FormFieldConfigDTO createCustomField(FormFieldConfigDTO dto) {
        String key = dto.getFieldKey();
        if (key == null || key.trim().isEmpty()) {
            key = "custom_" + dto.getFieldLabel().toLowerCase().replaceAll("[^a-z0-9]", "_");
        }
        key = key.trim();
        if (repository.existsByFieldKey(key)) {
            key = key + "_" + System.currentTimeMillis() % 10000;
        }

        FormFieldConfig entity = FormFieldConfig.builder()
                .fieldKey(key)
                .fieldLabel(dto.getFieldLabel())
                .tabName(dto.getTabName() != null ? dto.getTabName() : "Personal Info")
                .fieldType(dto.getFieldType() != null ? dto.getFieldType().toUpperCase() : "TEXT")
                .masterCategory(dto.getMasterCategory())
                .options(dto.getOptions())
                .isMandatory(Boolean.TRUE.equals(dto.getIsMandatory()))
                .isVisible(dto.getIsVisible() != null ? dto.getIsVisible() : true)
                .isCustom(true)
                .sortOrder(dto.getSortOrder() != null ? dto.getSortOrder() : 99)
                .placeholder(dto.getPlaceholder())
                .build();

        FormFieldConfig saved = repository.save(entity);
        log.info("Created custom form field: key={}, label={}", saved.getFieldKey(), saved.getFieldLabel());
        return FormFieldConfigDTO.fromEntity(saved);
    }

    @Transactional
    public void deleteCustomField(Long id) {
        FormFieldConfig config = repository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("FormFieldConfig not found with id: " + id));
        if (Boolean.TRUE.equals(config.getIsCustom())) {
            repository.delete(config);
            log.info("Deleted custom form field: {}", config.getFieldKey());
        } else {
            throw new IllegalArgumentException("Cannot delete standard system field: " + config.getFieldKey());
        }
    }
}

File: employee-management-ui/src/app/core/services/form-field-config.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { APIResponse } from '../models/api-response.model';
import { environment } from '../../../environments/environment';

export interface FormFieldConfig {
  id?: number;
  fieldKey: string;
  fieldLabel: string;
  tabName: string;
  fieldType: string; // 'TEXT' | 'NUMBER' | 'DATE' | 'SELECT' | 'BOOLEAN' | 'TEXTAREA'
  masterCategory?: string;
  options?: string;
  isMandatory: boolean;
  isVisible: boolean;
  isCustom: boolean;
  sortOrder: number;
  placeholder?: string;
  createdAt?: string;
  updatedAt?: string;
}

@Injectable({
  providedIn: 'root'
})
export class FormFieldConfigService {
  private baseUrl = environment.apiUrl;

  constructor(private http: HttpClient) {}

  getAllConfigs(): Observable<FormFieldConfig[]> {
    return this.http.get<APIResponse<FormFieldConfig[]>>(`${this.baseUrl}/form-fields`).pipe(
      map(res => res.data || [])
    );
  }

  getVisibleConfigs(): Observable<FormFieldConfig[]> {
    return this.http.get<APIResponse<FormFieldConfig[]>>(`${this.baseUrl}/form-fields/visible`).pipe(
      map(res => res.data || [])
    );
  }

  toggleMandatory(id: number): Observable<FormFieldConfig> {
    return this.http.put<APIResponse<FormFieldConfig>>(`${this.baseUrl}/form-fields/${id}/toggle-mandatory`, {}).pipe(
      map(res => res.data)
    );
  }

  toggleVisibility(id: number): Observable<FormFieldConfig> {
    return this.http.put<APIResponse<FormFieldConfig>>(`${this.baseUrl}/form-fields/${id}/toggle-visibility`, {}).pipe(
      map(res => res.data)
    );
  }

  updateBulk(configs: FormFieldConfig[]): Observable<FormFieldConfig[]> {
    return this.http.put<APIResponse<FormFieldConfig[]>>(`${this.baseUrl}/form-fields/bulk`, configs).pipe(
      map(res => res.data || [])
    );
  }

  createCustomField(config: Partial<FormFieldConfig>): Observable<FormFieldConfig> {
    return this.http.post<APIResponse<FormFieldConfig>>(`${this.baseUrl}/form-fields/custom`, config).pipe(
      map(res => res.data)
    );
  }

  deleteCustomField(id: number): Observable<void> {
    return this.http.delete<APIResponse<void>>(`${this.baseUrl}/form-fields/custom/${id}`).pipe(
      map(() => void 0)
    );
  }
}
